[REVIEW] Refactoring Quasi-Newton to use cumlHandle

yjk21 opened the pull request.

cjnolet commented:
add to whitelist

teju85 commented on cuML/src/glm/glm.cu:
```diff
@@ -180,29 +180,27 @@ void qnFit(float *X, float *y, int N, int D, int C, bool fit_intercept,
-  CUBLAS_CHECK(cublasCreate(&cublas_handle));
+  // TODO cuml will be passed in
+  cumlHandle cuml;
+  cuml.setStream(0);
```
This is no more needed after @vinaydes's fixes for initializing the stream during the ctor of `cumlHandle`.

teju85 commented on cuML/src/glm/glm.cu:
```diff
@@ -180,29 +180,27 @@ void qnFit(float *X, float *y, int N, int D, int C, bool fit_intercept,
         linesearch_max_iter, lbfgs_memory, verbosity, w0, f, num_iters,
-        X_col_major, loss_type, cublas_handle, stream);
-  CUBLAS_CHECK(cublasDestroy(cublas_handle));
+        X_col_major, loss_type, cuml.getImpl());
```
Minor knitpick: Similar to @jirikraus's comments [here](https://github.com/rapidsai/cuml/pull/482#pullrequestreview-227762358), please use the following interface to any functions inside cuML, while exposing `cumlHandle`:
1. `cumlHandle_impl` is the first argument to be passed
2. `cudaStream_t` should be passed explicitly as the last argument
If you need an example, dbscan clean-up by Vinay [here](https://github.com/rapidsai/cuml/blob/branch-0.7/cuML/src/dbscan/dbscan.h) should be helpful.

teju85 commented on cuML/src/glm/qn/qn.h:
```diff
@@ -42,61 +43,63 @@ int qn_fit(LossFunction &loss, T *Xptr, T *yptr, T *zptr, int N,
   STORAGE_ORDER ord = X_col_major ? COL_MAJOR : ROW_MAJOR;
 
-  SimpleMatOwning<T> z(C, N); // TODO this allocation somewhere else?
+  MLCommon::device_buffer<T> tmp(cuml.getDeviceAllocator(), cuml.getStream(),
```
please use an explicit `cudaStream_t` object instead of from the cumlHandle

teju85 commented on cuML/src/glm/qn/qn.h:
```diff
@@ -42,61 +43,63 @@ int qn_fit(LossFunction &loss, T *Xptr, T *yptr, T *zptr, int N,
   } break;
   default: { ASSERT(false, "qn.h: unknown loss function."); }
   }
+  tmp.release(cuml.getStream());
```
use an explicit `cudaStream_t`.

teju85 commented on cuML/src/glm/qn/qn_solvers.h:
```diff
@@ -317,21 +317,25 @@ inline OPT_RETCODE min_owlqn(const LBFGSParam<T> &param, Function &f,
   if (l1 == 0.0) {
-    SimpleVecOwning<T> workspace(lbfgs_workspace_size(opt_param, x.len));
+
+    MLCommon::device_buffer<T> tmp(cuml.getDeviceAllocator(), cuml.getStream(),
```
explicit stream

teju85 commented on cuML/src/glm/qn/qn_solvers.h:
```diff
@@ -317,21 +317,25 @@ inline OPT_RETCODE min_owlqn(const LBFGSParam<T> &param, Function &f,
                     num_iters, // output iterations
                     workspace, // scratch space
-                    stream, verbosity);
+                    cuml.getStream(), verbosity);
```
explicit stream

teju85 commented on cuML/src/glm/qn/qn_solvers.h:
```diff
@@ -317,21 +317,25 @@ inline OPT_RETCODE min_owlqn(const LBFGSParam<T> &param, Function &f,
                     workspace, // scratch space
-                    stream, verbosity);
+                    cuml.getStream(), verbosity);
+    tmp.release(cuml.getStream());
```
explicit stream

teju85 commented on cuML/src/glm/qn/qn_solvers.h:
```diff
@@ -343,7 +347,9 @@ inline int qn_minimize(SimpleVec<T> &x, T *fx, int *num_iters,
     // it needs to evaluate f(x) and its gradient separately
 
-    SimpleVecOwning<T> workspace(owlqn_workspace_size(opt_param, x.len));
+    MLCommon::device_buffer<T> tmp(cuml.getDeviceAllocator(), cuml.getStream(),
```
explicit stream

teju85 commented on cuML/src/glm/qn/qn_solvers.h:
```diff
@@ -352,7 +358,8 @@ inline int qn_minimize(SimpleVec<T> &x, T *fx, int *num_iters,
                     num_iters, // output iterations
                     workspace, // scratch space
-                    stream, verbosity);
+                    cuml.getStream(), verbosity);
```
explicit stream

teju85 commented on cuML/src/glm/qn/qn_solvers.h:
```diff
@@ -352,7 +358,8 @@ inline int qn_minimize(SimpleVec<T> &x, T *fx, int *num_iters,
                     workspace, // scratch space
-                    stream, verbosity);
+                    cuml.getStream(), verbosity);
+    tmp.release(cuml.getStream());
```
explicit stream

teju85 commented on cuML/src/glm/qn/simple_mat.h:
```diff
@@ -162,7 +162,7 @@ template <typename T> struct SimpleMat {
   }
 
-  inline void fill(const T val, cudaStream_t stream=0) {
+  inline void fill(const T val, cudaStream_t stream = 0) {
```
do not use default argument for stream.

teju85 commented on cuML/src/glm/qn/simple_mat.h:
```diff
@@ -245,8 +245,8 @@ inline T nrm2(const SimpleVec<T> &u, T *tmp_dev, cudaStream_t stream = 0) {
 template <typename T>
-inline T nrm1(const SimpleVec<T> &u, T *tmp_dev, cudaStream_t stream = 0){
-  MLCommon::LinAlg::rowNorm(tmp_dev, u.data,  u.len, 1, MLCommon::LinAlg::L1Norm,
+inline T nrm1(const SimpleVec<T> &u, T *tmp_dev, cudaStream_t stream = 0) {
```
do not use default arg for streams

teju85 commented on cuML/src/glm/qn/simple_mat.h:
```diff
@@ -245,8 +245,8 @@ inline T nrm2(const SimpleVec<T> &u, T *tmp_dev, cudaStream_t stream = 0) {
+  MLCommon::LinAlg::rowNorm(tmp_dev, u.data, u.len, 1, MLCommon::LinAlg::L1Norm,
                             true, stream, MLCommon::Nop<T>());
   T tmp_host;
   MLCommon::updateHost(&tmp_host, tmp_dev, 1);
```
I no this is not related to this PR, but please use `updateHostAsync` with an explicit stream

yjk21 replied:
Ok. I suppose then I would need to call `cudaStreamSynchronize` right after?

teju85 replied:
Yes. A stream sync needed right before when it gets consumed.
And ideally speaking, the `tmp_host` should be of type `host_buffer` and that way we could issue this async operation.

teju85 commented:
Tagging @jirikraus in case he too has any comments on this PR from interface POV.

yjk21 commented:
@teju85 Thanks a lot for the detailed review.

jirikraus commented:
> @teju85 Thanks a lot for the detailed review.

@yjk21 I will lake a look at this once you addressed @teju85 comments. Please tag me when you are don with that. Thanks Jiri

yjk21 commented:
@jirikraus I made the changes requested by @teju85. So whenever you're ready :)

Thanks guys

jirikraus commented on cuML/src/glm/glm.cu:
```diff
@@ -180,29 +180,25 @@ void qnFit(float *X, float *y, int N, int D, int C, bool fit_intercept,
-        X_col_major, loss_type, cublas_handle, stream);
-  CUBLAS_CHECK(cublasDestroy(cublas_handle));
+  // TODO cuml will be passed in
+  cumlHandle cuml;
```
Why is this a TODO in this PR? I thought the purpose of this PR is to introduce cumlHandle?

jirikraus commented on cuML/src/glm/qn/glm_base.h:
```diff
@@ -31,7 +31,7 @@ namespace GLM {
 template <typename T>
 inline void linearFwd(SimpleMat<T> &Z, const SimpleMat<T> &X,
-                      const SimpleMat<T> &W, const cublasHandle_t &cublas,
+                      const SimpleMat<T> &W, const cumlHandle_impl &cuml,
```
So far we used to pass in `cumlHandle_impl` has the first argument to every function taking it. I think it would be good to stick with that. As name mostly `handle` has used. This also applies to other functions in this PR I will not repeat the comment.

jirikraus commented on cuML/src/glm/qn/qn.h:
```diff
@@ -42,61 +42,63 @@ int qn_fit(LossFunction &loss, T *Xptr, T *yptr, T *zptr, int N,
-  SimpleMatOwning<T> z(C, N); // TODO this allocation somewhere else?
+  MLCommon::device_buffer<T> tmp(cuml.getDeviceAllocator(), stream, C * N);
+
+  SimpleMat<T> z(tmp.data(), C, N);
```
@teju85, @akkamesh Seems to be another occurrence of a tensor view class.

yjk21 replied:
Good point. 
We talked about this a while ago and decided to prioritize the features we are working on over unifying our abstractions, which also seemed non-trivial because they were designed with different use cases in mind.

I can imagine though that there might soon be a critical mass of code that could benefit from such abstractions. E.g. in the case of GLM, having the matrix concept made it relatively painless to transition to sparse data support.

jirikraus commented on cuML/src/glm/qn/qn.h:
```diff
@@ -42,61 +42,63 @@ int qn_fit(LossFunction &loss, T *Xptr, T *yptr, T *zptr, int N,
   } break;
   default: { ASSERT(false, "qn.h: unknown loss function."); }
   }
+  tmp.release(stream);
```
This is done implicitly when `tmp` get's out of scope. Any reason to do this explicitly? Also applies to other files.

jirikraus commented on cuML/src/glm/qn/simple_mat.h:
```diff
@@ -184,8 +184,8 @@ template <typename T> struct SimpleVec : SimpleMat<T> {
-                   cublasHandle_t &cublas) {
-    Super::assign_gemm(alpha, A, transA, x, false, beta, cublas);
+                   const cumlHandle_impl &cuml) {
+    Super::assign_gemm(alpha, A, transA, x, false, beta, cuml, 0);
```
Is the last argument (`0`) the stream? That should not be used.

jirikraus commented on cuML/src/glm/qn/simple_mat.h:
```diff
@@ -295,7 +297,8 @@ std::ostream &operator<<(std::ostream &os, const SimpleMat<T> &mat) {
 // TODO The following classes own their storage and are just for testing
-// The internals should be replaced by device_buffer+device_allocator, when ready
+// The internals should be replaced by device_buffer+device_allocator, when
+// ready
```
Shouldn't that be part of this PR?

teju85 commented:
Thanks @jirikraus for the feedback.

I'd like to add a point about exposing C API. With the amount of work we have, it's easier for everyone if we keep this as an optional, low-priority item unless there's a need for it.

jirikraus commented:
> I'd like to add a point about exposing C API. With the amount of work we have, it's easier for everyone if we keep this as an optional, low-priority item unless there's a need for it.

Agree it makes sense to consider this low-priority. However two arguments to consider for doing it anyway:
1. It is an important sanity check that the C++ API can indeed be easily accessed via a C-API. It is only a compile time check but better than only relying on reviewers.
2. It is relatively low effort.

yjk21 commented:
Thanks @jirikraus for the comments. 
I changed to comply with the naming/position convention, added the C API and implemented the owning mat/vec using device buffer.

I did not yet test if there are any remaining allocations/syncs when using RMM. 
Can you recommend a tool for this?

Thanks

jirikraus commented:
> I did not yet test if there are any remaining allocations/syncs when using RMM.
> Can you recommend a tool for this?

What work reasonably well for me is to do the following: Put a NVTX range[1] around the call to the entry function (e.g. `qnFit`) under observation then run with `nvprof` dumping a API trace with

```
nvprof --print-api-trace -o api_trace.log ./app <args>
```

In the API trace you can check the calls between the start and the end of the NVTX range. As an alternative to NVTX you could use `cudaProfilerStart` and `cudaProfilerStop` and use `nvprof` like this

```
nvprof --profile-from-start off --print-api-trace -o api_trace.log ./app <args>
```

Hope this helps

Jiri

[1] https://devblogs.nvidia.com/cuda-pro-tip-generate-custom-application-profile-timelines-nvtx/

jirikraus commented:
@yjk21 please tag me if it makes sense to take another look at this.

yjk21 commented:
Hi @jirikraus, thanks for the tips. Indeed, there were some synchronous dev2dev copies on the default stream. 
When you have time, I think it's ready for another look.
Thanks

dantegd commented:
rerun tests

jirikraus commented on cuML/src/glm/qn/glm_softmax.h:
```diff
@@ -186,11 +188,11 @@ void launchLogsoftmax(T *loss_val, T *dldZ, const T *Z, const T *labels, int C,
+  inline void getLossAndDZ(T *loss_val, SimpleMat<T> &Z, const SimpleVec<T> &y,
+                           cudaStream_t stream) {
 
     launchLogsoftmax(loss_val, Z.data, Z.data, y.data, Z.m, Z.n, stream);
```
launchLogsoftmax calls synchronous `cudaMemset` which due to default stream synchronization implicitly synchronizes all streams. This should be changed. Why was that missed in your tests?

yjk21 replied:
Good question.
I wrote separate test code to play around with RMM and forgot to add it.

yjk21 replied:
I ran the analysis on the unit test. Within the relevant range, there is no more activity on the default stream.

jirikraus commented on cuML/src/glm/qn/simple_mat.h:
```diff
@@ -162,41 +163,39 @@ template <typename T> struct SimpleMat {
-                          cudaMemcpyDeviceToDevice));
+    CUDA_CHECK(cudaMemcpyAsync(data, other.data, len * sizeof(T),
+                               cudaMemcpyDeviceToDevice, stream));
+    CUDA_CHECK(cudaStreamSynchronize(stream));
```
`copy` takes a stream as an argument. From a callersperspetive this expresses that it is an asynchronous operation. Please remove the `cudaStreamSynchronize`. This is something the caller want's to control. If you want to be more explicit about this add an async suffix to copy.

yjk21 replied:
Good point!

jirikraus commented on cuML/src/glm/qn/simple_mat.h:
```diff
@@ -294,55 +294,48 @@ std::ostream &operator<<(std::ostream &os, const SimpleMat<T> &mat) {
-    Super::reset(this->data, n);
-    allocated = true;
+  void reset(const cumlHandle_impl &handle, int n, cudaStream_t stream) {
+    buf.reset(new Buffer(handle.getDeviceAllocator(), stream, n));
```
Could you simply call `buf->resize(n,stream)` instead?

yjk21 replied:
Do you have a strong preference for this?
This would mean that the default constructor of `SimpleMatOwning` need the allocator as well to create an empty buffer object. As it is, it might be clearer, when the actual allocation happens.

jirikraus replied:
As it probably makes sense to rework the different tensor/matrix abstractions to be generally usable across different cuML algorithms some time I do not have a strong preference here. Just to explain my thoughts: I think it makes sense that a class that owns memory (and should use the provided allocator infrastructure for that) needs an allocator object to get constructed. So my preference would be do delete the default constructor.  Does that make sense to you?

yjk21 replied:
Yeah, makes a lot of sense. I removed the default constructor

jirikraus commented on cuML/src/glm/qn/simple_mat.h:
```diff
@@ -294,55 +294,48 @@ std::ostream &operator<<(std::ostream &os, const SimpleMat<T> &mat) {
-    Super::reset(Super::data, m, n);
-    allocated = true;
+  void reset(const cumlHandle_impl &handle, int m, int n, cudaStream_t stream) {
+    buf.reset(new Buffer(handle.getDeviceAllocator(), stream, m * n));
```
Could you simply call `buf->resize(n,stream)` instead?

yjk21 replied:
Reset was removed

jirikraus commented on cuML/test/quasi_newton.cu:
```diff
@@ -20,11 +21,14 @@ struct QuasiNewtonTest : ::testing::Test {
-    cublasCreate(&cublas);
-    stream = 0;
+    cumlError_t status;
+    cumlCreate(&cuml_handle);
```
Why us the C instead of the C++ API here? Using the C++ API would simplify the code.

jirikraus commented on cuML/src/glm/qn/simple_mat.h:
```diff
@@ -294,55 +294,48 @@ std::ostream &operator<<(std::ostream &os, const SimpleMat<T> &mat) {
-  SimpleVecOwning(int n) : Super(), allocated(false) { reset(n); }
-
-  ~SimpleVecOwning() noexcept(false) { CUDA_CHECK(cudaFree(Super::data)); }
+  SimpleVecOwning(const cumlHandle_impl &handle, int n, cudaStream_t stream)
```
Why not only pass in a `std::shared_ptr<device_allocator>` instead of `cumlHandle_impl`?

jirikraus commented on cuML/src/glm/qn/simple_mat.h:
```diff
@@ -294,55 +294,48 @@ std::ostream &operator<<(std::ostream &os, const SimpleMat<T> &mat) {
-  SimpleMatOwning(int m, int n, STORAGE_ORDER order = COL_MAJOR)
-      : Super(order), allocated(false) {
-    reset(m, n);
+  SimpleMatOwning(const cumlHandle_impl &handle, int m, int n,
```
Why not only pass in a `std::shared_ptr<device_allocator>` instead of `cumlHandle_impl`?

jirikraus commented on cuML/test/quasi_newton.cu:
```diff
@@ -45,17 +49,17 @@ const double QuasiNewtonTest::X[QuasiNewtonTest::N][QuasiNewtonTest::D] = {
 
-  SimpleVecOwning<T> w_ref(C * (D + fit_intercept));
+  SimpleVecOwning<T> w_ref(handle, C * (D + fit_intercept), handle.getStream());
   updateDevice(w_ref.data, &w_ref_cm[0], C * D);
```
Just pointing out that `updateDevice` possible relies on default stream semantics, i.e. that it implicitly syncs with `handle.getStream()`. This will break if the stream passed into `handle` is created with the non blocking flag. This also applies to other parts of the test.

jirikraus commented on cuML/test/quasi_newton.cu:
```diff
@@ -96,17 +104,19 @@ T run(LossFunction &loss, DevUpload<T> &devUpload, InputSpec &in, T l1, T l2,
+T run_api(const cumlHandle &cuml_handle, int loss_type, int C,
+          bool fit_intercept, DevUpload<T> &devUpload, InputSpec &in, T l1,
+          T l2, T *w, SimpleMat<T> &z, int verbosity = 0,
           cudaStream_t stream = 0) {
```
I think it woudl be better to remove the default of `0` for stream.

jirikraus commented on cuML/test/quasi_newton.cu:
```diff
@@ -136,13 +148,14 @@ TEST_F(QuasiNewtonTest, binary_logistic_vs_sklearn) {
 
-  DevUpload<double> devUpload(in, &X[0][0], &y[0], cublas);
+  DevUpload<double> devUpload(handle, in, &X[0][0], &y[0],
+                              handle.getCublasHandle());
```
Why does `devUpload` take `handle.getCublasHandle()` and `handle` as arguments?

jirikraus commented on cuML/test/quasi_newton.cu:
```diff
@@ -152,12 +165,13 @@ TEST_F(QuasiNewtonTest, binary_logistic_vs_sklearn) {
   l1 = alpha;
   l2 = 0.0;
-  fx = run(loss_b, devUpload, in, l1, l2, w0.data, z, 0, stream);
+  fx = run(loss_b, devUpload, in, l1, l2, w0.data, z, handle, 0);
```
Is there a sync needed after run to make sure `compApprox` works correctly?

yjk21 replied:
The sync happens everytime the optimizer evaluates the objective to get fx to the host, where the control flow happens.

jirikraus commented:
@yjk21 please tag me when it makes sense to take another look.

yjk21 commented on cuML/src/glm/qn/glm_base.h:
```diff
@@ -169,7 +164,10 @@ template <typename T, class GLMObjective> struct GLMWithData : GLMDims {
     Mat G(gradFlat.data, C, dims);
```
@jirikraus `SimpleMat` and `SimpleVec` are simple views on provided gpu memories. The only point in `glm/qn` folder where a mem alloc could happen is in the entry point in the API function in `qn.h` where the complete workspace is created/requested.

The conversion between Vec and Mat is just a wrapping for semantic reasons, because conceptually the non-linear solver optimizes functions on R^D, while the GLM loss functions are modelled to have parameter matrices (for the case of multi-task/multi-class problems).

yjk21 commented on cuML/src/glm/qn/glm_base.h:
```diff
@@ -48,16 +48,16 @@ inline void linearFwd(SimpleMat<T> &Z, const SimpleMat<T> &X,
     MLCommon::LinAlg::matrixVectorOp(Z.data, Z.data, bias.data, Z.n, Z.m, false,
```
Yes, that's correct.

yjk21 commented on cuML/src/glm/qn/glm_base.h:
```diff
@@ -193,14 +192,16 @@ void numeric_grad(Loss &loss, const T *X, const T *y, const T *w,
-    modKernel<<<MLCommon::ceildiv(len, 256), 256, 0, stream>>>(w_mod.data, d, h);
+    modKernel<<<MLCommon::ceildiv(len, 256), 256, 0, stream>>>(w_mod.data, d,
+                                                               h);
     cudaThreadSynchronize();
```
Ok. I'll take the function out of this PR and submit it separately, rethinking where it should live.

yjk21 commented on cuML/src/glm/qn/glm_base.h:
```diff
@@ -193,14 +192,16 @@ void numeric_grad(Loss &loss, const T *X, const T *y, const T *w,
-    modKernel<<<MLCommon::ceildiv(len, 256), 256, 0, stream>>>(w_mod.data, d, -2 * h);
+    modKernel<<<MLCommon::ceildiv(len, 256), 256, 0, stream>>>(w_mod.data, d,
+                                                               -2 * h);
     cudaThreadSynchronize();
```
Ok. I'll take the function out of this PR and submit it separately, rethinking where it should live.

yjk21 commented on cuML/src/glm/qn/glm_base.h:
```diff
@@ -193,14 +192,16 @@ void numeric_grad(Loss &loss, const T *X, const T *y, const T *w,
     CUDA_CHECK(
         cudaMemcpy(w_mod.data, w, len * sizeof(T), cudaMemcpyDeviceToDevice));
```
Ok. I'll take the function out of this PR and submit it separately, rethinking where it should live.

yjk21 commented:
@jirikraus sorry for taking so much of your time with the many iterations. I commited a couple more changes.
Thanks

jirikraus commented on cuML/src/glm/qn/glm_regularizer.h:
```diff
@@ -58,12 +59,19 @@ template <typename T, class Loss, class Reg> struct RegularizedGLM : GLMDims {
+
+    T reg_host; 
+    MLCommon::updateHostAsync(&reg_host, lossVal.data, 1, stream);
+    CUDA_CHECK(cudaStreamSynchronize(stream));
```
`reg_host` is not used before line 74 so one `cudaStreamSynchronize` in line 72 is sufficient and the `cudaStreamSynchronize` in line 67 is unnecessary.

jirikraus commented on cuML/src/glm/qn/qn_solvers.h:
```diff
@@ -138,8 +138,9 @@ inline OPT_RETCODE min_lbfgs(const LBFGSParam<T> &param,
-    gradp = grad;
+    xp.copy_async(x, stream);
+    gradp.copy_async(grad, stream);
+    CUDA_CHECK(cudaStreamSynchronize(stream));
```
are the `copy_async` of `x` to `xp` and `grad` to `gradp` device to device copies? If so I think the `cudaStreamSynchronize` is not really needed as subsequent operations are submitted in the same stream, right?

yjk21 replied:
Yes, it's `DeviceToDevice`. 
I think you're right that we can remove the syncs pointed out here and below.
Sync points are in principle quite well defined, i.e. whenever a scalar is computed and transferred to the host. 
I guess I have to get a bit more familiar with the whole stream mechanics

jirikraus commented on cuML/src/glm/qn/qn_solvers.h:
```diff
@@ -175,7 +176,8 @@ inline void update_pseudo(const SimpleVec<T> &x, const SimpleVec<T> &grad,
   if (grad.len > pg_limit) {
-    pseudo = grad;
+    pseudo.copy_async(grad, stream);
+    CUDA_CHECK(cudaStreamSynchronize(stream));
```
Is the `copy_async` of `grad` to `pseudo` a device to device copy? If so I think the `cudaStreamSynchronize` is not really needed as subsequent operations are submitted in the same stream, right?

yjk21 replied:
See above

jirikraus commented on cuML/src/glm/qn/qn_solvers.h:
```diff
@@ -275,8 +277,9 @@ inline OPT_RETCODE min_owlqn(const LBFGSParam<T> &param, Function &f,
-    gradp = grad;
+    xp.copy_async(x, stream);
+    gradp.copy_async(grad, stream);
+    CUDA_CHECK(cudaStreamSynchronize(stream));
```
See my comment on line 143 is this `cudaStreamSynchronize` necessary?

yjk21 replied:
See above

jirikraus commented on cuML/src/glm/qn/simple_mat.h:
```diff
@@ -294,55 +292,40 @@ std::ostream &operator<<(std::ostream &os, const SimpleMat<T> &mat) {
   typedef SimpleVec<T> Super;
-  bool allocated;
+  typedef MLCommon::device_buffer<T> Buffer;
+  std::unique_ptr<Buffer> buf;
```
Does this need to be a `unique_ptr` or could this be declared as ` Buffer buf;`

yjk21 commented:
Hi @cjnolet and @oyilmaz-nvidia,
do you think we can move forward with this PR?

Thanks

cjnolet commented:
@yjk21, I believe @dantegd is in need of a predict() function so that he can finish the cython wrapper. Do you think you will be able to do that by Friday (final burn down for 0.7)?

yjk21 commented:
Hey @cjnolet, yeah I think I can get that in before your sod on Friday.
Thanks

yjk21 commented:
@dantegd, @cjnolet I added the predict methods
